### ISSUE TITLE
EarlyStopping with GridSearchCV

### ISSUE THREAD

**Issue by ilkersigirci** [skorch-dev/skorch#837]
I am trying to construct a pipeline that cross validates pytorch models with sklearn GridSearchCV by using callbacks like EarlyStopping. I skim through docs and issues but couldn't find related information about it. Is this currently not supported?

I know I can use callbacks when NeuralNetRegressor with following configuration:
```python
net_regr = NeuralNetRegressor(
    RegressorModule,
    max_epochs=10,
    lr=0.1,
    device='cpu',
    criterion=torch.nn.MSELoss,
    batch_size=32,
    train_split=ValidSplit(5),
    iterator_train__shuffle=False,
    iterator_train__num_workers=4,
    optimizer=torch.optim.Adam,
    callbacks=[early_stopping, lrscheduler]
)

net_regr.fit(X_regr, y_regr)
```

But if I want to do cross validation with GridSearchCV, I should set NeuralNetRegressor.train_split to None in order to not do double cv. Unfortunately, following configuration doesn't work, which raises the error `KeyError: "Key 'valid_loss' was not found in history."`.
```python
net_regr = NeuralNetRegressor(
    RegressorModule,
    max_epochs=10,
    lr=0.1,
    device='cpu',
    criterion=torch.nn.MSELoss,
    batch_size=32,
    train_split=None,
    iterator_train__shuffle=False,
    iterator_train__num_workers=4,
    optimizer=torch.optim.Adam,
    callbacks=[early_stopping, lrscheduler]
)

params = {
    # 'lr': [0.05, 0.1],
    'module__num_units': [10, 20],
}

gs = GridSearchCV(net_regr, params, refit=False, cv=TimeSeriesSplit(), scoring='r2', verbose=2)
gs.fit(X_regr, y_regr)
```

**Comment by BenjaminBossan** [comment]
> But if I want to do cross validation with GridSearchCV, I should set NeuralNetRegressor.train_split to None in order to not do double cv.

This is correct. However, for early stopping to work, you need the validation loss. And this requires there to be a validation set. Thus if you want to use early stopping like this, you need to keep the train split, even if it means that some data is "wasted".

Alternatively, you could early stop not on the `'valid_loss'` but on something else based on the training dataset. E.g. you could monitor the train loss: `EarlyStopping(patience=5, monitor='train_loss')`. It's up to you to decide which option is better for you.

**Comment by ilkersigirci** [comment]
Alright, I thought the same. I just didn't want to waste the data. I will use `train_loss` for my experiments, then. Thanks

**Comment by franxav79** [comment]
I have the same issue, GridSearchCV and TimeSeriesSplit, how I can recover the valid_loss from gs and see how well the model train.

Or the aproach is to take the best parameters result and to test in validation set by separate.
I am not using call backs.

net = NeuralNet(
    module=LSTM_E_1, max_epochs=epochs,
    optimizer=torch.optim.Adam,
    criterion=criterion,
    lr=lr, verbose=1, 
    device=device, 
    )

params = {
'lr': [1, 0.1, 0.01, 0.001, 0.0001]}

net.set_params(verbose=0, train_split= None)
tscv = TimeSeriesSplit(n_splits=6, test_size=2)
gs = GridSearchCV(net, 
                    params, 
                    refit=True, 
                    scoring='r2', 
                    verbose=1, cv=tscv, 
                    #return_valid_score=True
                    )
